cmake/std/trilinosprhelpers/setenvironment/SetEnvironment.py

```diff
@@ -0,0 +1,447 @@
+        print("Environment Vars")
+        print("+--------------+")
+        for envvar in self.actions["setenv"]:
+            # Note: we can't print the expansion here because the source
+            #       envvar might not exist yet.
+            k = envvar['key']
+            v = envvar['value']
+            print("[envvar] export {}={}".format(k,v))
+        for k in self.actions["unsetenv"]:
+            print("[envvar] unset {}".format(k))
+
+        return 0
+
+
+    def _expand_envvars_in_string(self, string_in):
+        """
+        Take an input string that may contain environment variables in the style
+        of BASH shell environment vars (i.e., "${foobar}") and replace them with
+        the actual environment variables.
+
+        Returns:
+            A string that contains the contents of any `${ENVVAR}` entries expanded
+            inline into the string.
+
+        Raises:
+            KeyError: Required environment variable does not exist.
+        """
+        regexp = re.compile(r"(\$\{(\S*)\})")
+        string_out = string_in
+        for m in re.finditer(regexp, string_out):
+            #v = m.group(1)  # The full ENVVAR sequence: ${VARNAME}
+            s = m.group(2)  # Just the ENVVAR itself: VARNAME
+            if(s in os.environ.keys()):
+                string_out = re.sub(regexp, os.environ[s], string_in)
+            else:
+                msg = "Required environment variable `{}` does not exist.".format(s)
+                raise KeyError(msg)
+        return string_out
+
+
+    def _load_configuration(self):
+        """
+        Loads the environment data from the provided `profile` and saves the
+        information to the `actions` property.
+
+        This is a recursive operation (recurses on the `use <profile>:` entries
+        in the configuration file). We use a DFS recursion strategy with back-edge
+        checking to prevent cyclic recursion. We do not
+
+        Returns:
+            A dict containing the contents of self.actions.
+        """
+        self.actions = { "setenv": [],         # Envvars that we'll set
+                         "unsetenv": [],       # Envvars to explicitly unset (after setting)
+                         "module-op": [],      # module operations (use, load, unload, etc.)
+                         "module-list": {}     # Keys = modules already set to 'load'
+                        }
+
+        config = self.config
+
+        self.actions = self._load_configuration_r(config, self.profile, actions=self.actions)
+
+        return self.actions
+
+
+    def _load_configuration_r(self, config, profile, actions=None, processed_secs=None):
+        """
+        Recursive handler for _load_configuration.
+        Recursion is called when the operation is a 'use' operation in the key.
+
+        Raises:
+            KeyError: If `profile` section is not contained in `config`
+            Exception: If a section isn't loaded for a reason other than a KeyError.
+        """
```

Comment:
This routine is a little tough to follow.  I'm sure it makes good sense to you, as you're intimately familiar with the `*.ini` file syntax and what all that means, but as someone new coming up to speed on it, it's hard to tell what the various keys/values really mean wherever they're used.

Reply:
This is the internal recursive call... this is the part that actually does the parsing of the .ini file.  

I updated the docstring a little and cleaned up a few things. Turns out the `config` parameter isn't necessary as a parameter since it's the same as the property `self.config`, so that simplifies things a bit.  I renamed the profile parameter to be `config_section` to better describe that this corresponds to a section in the .ini file `[section name]` in a config file.

We could get into the details of this how what this does but it's probably not easy to do in a PR comment.